code/go/0chain.net/smartcontract/zcnsc/stakepool.go

```diff
@@ -0,0 +1,313 @@
+func (sp *StakePool) Decode(input []byte) error {
+	return json.Unmarshal(input, sp)
+}
+
+// save the stake pool
+func (sp *StakePool) save(sscKey, providerID string, balances cstate.StateContextI) (err error) {
+	_, err = balances.InsertTrieNode(stakePoolKey(sscKey, providerID), sp)
+	return
+}
+
+// The stake() returns total stake size including delegate pools want to unstake.
+func (sp *StakePool) stake() (stake state.Balance) {
+	for _, dp := range sp.Pools {
+		stake += dp.Balance
+	}
+	return
+}
+
+// empty a delegate pool if possible, call update before the empty
+func (sp *StakePool) empty(sscID, poolID, clientID string, balances cstate.StateContextI) (bool, error) {
+	var dp, ok = sp.Pools[poolID]
+	if !ok {
+		return false, fmt.Errorf("no such delegate pool: %q", poolID)
+	}
+
+	if dp.DelegateID != clientID {
+		return false, errors.New("trying to unlock not by delegate pool owner")
+	}
+
+	transfer := state.NewTransfer(sscID, clientID, dp.Balance)
+	if err := balances.AddTransfer(transfer); err != nil {
+		return false, err
+	}
+
+	sp.Pools[poolID].Balance = 0
+	sp.Pools[poolID].Status = spenum.Deleting
+
+	return true, nil
+}
+
+//
+// smart contract methods
+//
+
+// getStakePool of given authorizer
+func (zcn *ZCNSmartContract) getStakePool(authorizerID datastore.Key, balances cstate.StateContextI) (sp *StakePool, err error) {
+	sp = newStakePool()
+	err = balances.GetTrieNode(stakePoolKey(zcn.ID, authorizerID), sp)
+	if err != nil {
+		return nil, err
+	}
+
+	return sp, nil
+}
+
+// initial or successive method should be used by add_authorizer
+
+// SC functions
+
+// get existing stake pool or create new one not saving it
+func (zcn *ZCNSmartContract) getOrUpdateStakePool(
+	gn *GlobalNode,
+	authorizerId datastore.Key,
+	providerType spenum.Provider,
+	settings stakepool.StakePoolSettings,
+	balances cstate.StateContextI,
+) (*StakePool, error) {
+	if err := validateStakePoolSettings(settings, gn); err != nil {
+		return nil, fmt.Errorf("invalid stake_pool settings: %v", err)
+	}
+
+	changed := false
+
+	// the stake pool can be created by related validator
+	sp, err := zcn.getStakePool(authorizerId, balances)
+	if err != nil {
+		if err != util.ErrValueNotPresent {
+			return nil, fmt.Errorf("unexpected error: %v", err)
+		}
+		sp = newStakePool()
+		sp.Minter = cstate.MinterStorage
+		sp.Settings.DelegateWallet = settings.DelegateWallet
+		changed = true
+	}
+
+	if sp.Settings.MinStake != settings.MinStake {
+		sp.Settings.MinStake = settings.MinStake
+		changed = true
+	}
+
+	if sp.Settings.MaxStake != settings.MaxStake {
+		sp.Settings.MaxStake = settings.MaxStake
+		changed = true
+	}
+
+	if sp.Settings.ServiceCharge != settings.ServiceCharge {
+		sp.Settings.ServiceCharge = settings.ServiceCharge
+		changed = true
+	}
+
+	if sp.Settings.MaxNumDelegates != settings.MaxNumDelegates {
+		sp.Settings.MaxNumDelegates = settings.MaxNumDelegates
+		changed = true
+	}
+
+	if changed {
+		return sp, nil
+	} else {
```

Comment:
avoid `if .. else..`, use the following style instead:

```go
if changed {
    return sp, nil
}

return nil, fmt.Errorf("...")
```

Reply:
updated